CARDS-2017 / CARDS-2019 - Add a `draftLifetime` field to the `Patient Access` configuration

marta- opened the pull request:
See parent task [CARDS-2017 - _As an admin, I can configure for how long the patients' draft responses are kept in the database_](https://phenotips.atlassian.net/browse/CARDS-2017) on jira for context.

**Status**
* This PR is a DRAFT because the branch is based on #1310, which addresses issues related to resetting to initial values, that were particularly prominent after the addition of the new configuration field to PatientAccess (in particular, resetting a config field to an empty value)
* The newly introduced configuration field will be leveraged by #1298

**Testing**
* The added configuration is not being used anywhere yet on this branch, so testing should focus exclusively on managing its value in the administration
* As admin, navigate to Administration > Patient access
  * Note the new field labeled "`Patients can edit unsubmitted responses for:"
  * Check that if unset, its placeholder is -1
  * Check if the helper text under the input is indeed helping
  * Test that it can be modified, saved and retrieved in the administration

marta- committed:
CARDS-2017 / CARDS-2019 - Add a `draftLifetime` field to the `PatientAccess` configuration

sashaandjic reviewed (approved):
tested

![image](https://user-images.githubusercontent.com/23243244/213847515-a356d234-b97c-4db8-a313-be62cf076bec.png)

sdumitriu reviewed (commented):
I'm not sure defaulting to `allowedPostVisitCompletionTime` is right. In proms, that is `0`, and it's not the same as "the token's lifetime". We send an email 7 days before, the patient starts filling in, but since the "post visit completion time" is 0, their answers will be deleted at the end of the day, not 7 days later when the visit happens and the invitation expires. I would say it should default to `-1`, which means that draft responses are not deleted until the whole invitation expires, which is also when the whole forms will be deleted by the other cleanup task.

- A value of `-1` means no draft answers deletion until the whole form is deleted; this special value basically disables the draft answer deletion task
- A value of `0` means incomplete answers are deleted at midnight, but this comes with the risk that if the patient starts filling in at 23:55, some of his answers will be deleted 5 minutes later
- A value of `2` means that we let the answers live for 2 days, or when the invitation expires, whichever comes first

WDYT?

marta- commented:
> I'm not sure defaulting to `allowedPostVisitCompletionTime` is right. In proms, that is `0`, and it's not the same as "the token's lifetime". We send an email 7 days before, the patient starts filling in, but since the "post visit completion time" is 0, their answers will be deleted at the end of the day, not 7 days later when the visit happens and the invitation expires. I would say it should default to `-1`, which means that draft responses are not deleted until the whole invitation expires, which is also when the whole forms will be deleted by the other cleanup task.
> 
> * A value of `-1` means no draft answers deletion until the whole form is deleted; this special value basically disables the draft answer deletion task
> * A value of `0` means incomplete answers are deleted at midnight, but this comes with the risk that if the patient starts filling in at 23:55, some of his answers will be deleted 5 minutes later
> * A value of `2` means that we let the answers live for 2 days, or when the invitation expires, whichever comes first
> 
> WDYT?

I agree that defaulting to `allowedPostVisitCompletionTime` no longer fits the new specs of the cleanup task. Will update.

marta- committed:
CARDS-2017 / CARDS-2019 - Add a `draftLifetime` field to the `PatientAccess` configuration
* Updated the default value and associated comments
* Adjusted styling on the Patient Access admin page

marta- commented:
> > I'm not sure defaulting to `allowedPostVisitCompletionTime` is right. In proms, that is `0`, and it's not the same as "the token's lifetime". We send an email 7 days before, the patient starts filling in, but since the "post visit completion time" is 0, their answers will be deleted at the end of the day, not 7 days later when the visit happens and the invitation expires. I would say it should default to `-1`, which means that draft responses are not deleted until the whole invitation expires, which is also when the whole forms will be deleted by the other cleanup task.
> > 
> > * A value of `-1` means no draft answers deletion until the whole form is deleted; this special value basically disables the draft answer deletion task
> > * A value of `0` means incomplete answers are deleted at midnight, but this comes with the risk that if the patient starts filling in at 23:55, some of his answers will be deleted 5 minutes later
> > * A value of `2` means that we let the answers live for 2 days, or when the invitation expires, whichever comes first
> > 
> > WDYT?
> 
> I agree that defaulting to `allowedPostVisitCompletionTime` no longer fits the new specs of the cleanup task. Will update.

Updated the jira specs, the code (commit 9c87c79) and the PR description to reflect the above. Also rebased on the latest `dev`.

sdumitriu commented on modules/patient-portal/src/main/java/io/uhndata/cards/patients/api/PatientAccessConfiguration.java:
```diff
@@ -49,4 +49,14 @@
+     *
+     * @return A number of days
+     */
+    int getDraftLifetime();
```
I would update the description to mention the special meaning of `-1`.

marta- replied:
Updated in b76f27b.

marta- committed:
CARDS-2017 / CARDS-2019 - Add a `draftLifetime` field to the `PatientAccess` configuration
Addressing review comments: javadoc improvement

marta- committed:
CARDS-2017 / CARDS-2019 - Add a `draftLifetime` field to the `PatientAccess` configuration
Addressing review comments: put limits to the values the user can specify for draftLifetime
Also removed unused custom placeholder

sashaandjic commented:
Should we check if the value entered is within a valid range (e.g -1 ; 30 )?
![image](https://user-images.githubusercontent.com/23243244/214221176-33a07a75-51ab-4f11-b767-ce2991e625b3.png)

marta- commented:
> Should we check if the value entered is within a valid range (e.g -1 ; 30 )?

Added very basic min/max restrictions via html input attributes in 3358b3b. We can improve usability in the next iteration.

sashaandjic reviewed (approved):
tested
![image](https://user-images.githubusercontent.com/23243244/214453115-45144bf2-f642-4b46-a92a-96ba26fce371.png)

veronikaslc commented:
I prefer input validation `onChange()` to give an immediate feedback rather than make user click save button to see that the field is invalid

veronikaslc commented:
Clicking on `Reset to initial settings` 
- does not reset the second input (known bug, but still)
- saves any invalid input

marta- commented:
> Clicking on `Reset to initial settings`
> 
> * does not reset the second input (known bug, but still)
> * saves any invalid input

Fair enough, can you fix it @veronikaslc ?

veronikaslc commented:
> Clicking on `Reset to initial settings`
> 
> * does not reset the second input (known bug, but still)
> * saves any invalid input

@marta- wdyt about disabling save button and validating input before save click?

veronikaslc commented:
> I prefer input validation `onChange()` to give an immediate feedback rather than make user click save button to see that the field is invalid

@marta- wdyt about disabling save button and validating input before save click?

veronikaslc committed:
CARDS-2017 / CARDS-2019 - Add a  field to the  configuration
Fixed reset to initial values

marta- commented on modules/patient-portal/src/main/frontend/src/patient-portal/PatientAccessConfiguration.jsx:
```diff
@@ -64,16 +67,29 @@ function PatientAccessConfiguration() {
     allowedPostVisitCompletionTime: ["Patients can fill out surveys after the associated event for:"],
     draftLifetime: ["Patients can edit unsubmitted responses for:", "-1 means that drafts are kept until the patient is no longer able to access their surveys"]
   };
+  const errorText = `Value must be between ${DEFAULT_PATIENT_ACCESS_CONFIG['draftLifetime']} and ${patientAccessConfig?.allowedPostVisitCompletionTime}`;
```
`DEFAULT_PATIENT_ACCESS_CONFIG['draftLifetime']` is not necessarily the lowest allowed value. It happens to be 0 now, but if for any reason we had decided to make the default 3, we should still allow to configure more frequent cleanups. Please hardcode 0 as the lowest allowed value.

marta- replied:
Nevermind, I was misremembering the default we're using and its meaning. The implementation is correct, but the message should say something along the lines of:
```suggestion
  const errorText = `Please use a value between 0 and ${patientAccessConfig?.allowedPostVisitCompletionTime}, or ${DEFAULT_PATIENT_ACCESS_CONFIG['draftLifetime']} to disable periodic draft deletion.`;
```

marta- commented on modules/patient-portal/src/main/frontend/src/patient-portal/PatientAccessConfiguration.jsx:
```diff
@@ -53,22 +55,41 @@ function PatientAccessConfiguration() {
   };
+  const errorText = `Value must be between ${DEFAULT_PATIENT_ACCESS_CONFIG['draftLifetime']} and ${patientAccessConfig?.allowedPostVisitCompletionTime}`;
+
+  let readpatientAccessConfigData = (json) => {
```
Please fix inconsistent casing:

```suggestion
  let readPatientAccessConfigData = (json) => {
```

marta- commented on modules/data-entry/src/main/frontend/src/questionnaireEditor/MarkdownText.jsx:
```diff
@@ -46,6 +46,10 @@ let MarkdownText = (props) => {
   cmd.push(commands.divider);
   cmd.push(infoButton);
 
+  useEffect(() => {
+    setValue(props.value || '');
+  }, [props.value]);
+
```
How is this change related to the 2019 task?

veronikaslc replied:
It doesn't, thats why it's in a separate commit with a separate message

marta- replied:
Better than in the same commit, but when the change really has nothing at all to do with the current task and is even in a completely separate file, the best practice is to put it in another [misc] PR.

veronikaslc replied:
Separated to the PR https://github.com/data-team-uhn/cards/pull/1308

veronikaslc committed:
CARDS-2017 / CARDS-2019 - Add a  field to the  configuration
Addressed codereview comments: fixed casing, hardcoded the draftLifetime lower limit

veronikaslc committed:
CARDS-2017 / CARDS-2019 - Add a  field to the  configuration
Addressed codereview comments: better draftLifetime invalid error message

veronikaslc commented:
Rebased on the latest `dev`

marta- commented:
Looks like we're operating under deprecated specs. The draft lifetime should no longer be limited by token lifetime based on how we use it. One scenario that works poorly with this limitation is the PROMs one, where responses are due the day of the visit (so `allowedPostVisitCompletionTime` is 0, but we must delete drafts after 24h (the day after at midnight), do `draftLifetime` should be 1.

Also, I'm finding the implementation of `draftLifetime` with its own rendering function a bit too custom, and I'm not understanding the purpose / the value brought by using the `inputRef`.

marta- committed:
CARDS-2017 / CARDS-2019 - Add a `draftLifetime` field to the `PatientAccess` configuration
* Modified the logic to comply with updated specs: the draftLiftime value doesn't need to be less than or equal to allowedPostVisitCompletionTime
* Made the code that deals with out of bound values in inputs more generic
* Improved labels and helper texts
* Capitalized constant names per code style guidelines

veronikaslc commented:
The inputRef allows control over the input value as it is not updated on the reset action. Probably there are easier ways but I found thise workaround for the proflem.

marta- commented:
@veronikaslc please check that I haven't introduced any regressions to your intended behavior in the last commit.

marta- commented:
> The inputRef allows control over the input value as it is not updated on the reset action. Probably there are easier ways but I found thise workaround for the proflem.

I see, there is a regression then.

marta- committed:
[misc] Prevent the default values from being saved for token lifetime and draft lifetime in the PatientAccess admin screen

marta- commented:
Rebased on #1310 which separately addresses issues with resetting the data to empty values across all admin config screens.

sashaandjic reviewed (approved):
tested
